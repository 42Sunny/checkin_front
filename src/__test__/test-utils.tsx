// test-utils.jsx
import { configureStore } from "@reduxjs/toolkit";
import { Queries, render, RenderResult } from "@testing-library/react";
import { createMemoryHistory } from "history";
import React, { ReactElement } from "react";
import { Provider } from "react-redux";
import { Router } from "react-router-dom";
// Import your own reducer
import renderer from "react-test-renderer";
import appStore from "../redux/configureStore";
import rootReducer, { RootState } from "../redux/modules";
import { initialState as clusterInitState } from "../redux/modules/cluster";
import { initialState as userInitState } from "../redux/modules/user";

const makeCustomStore = ({
  cluster,
  user,
}: {
  cluster?: typeof clusterInitState;
  user?: typeof userInitState;
} = {}) => {
  return {
    clusterReducer: { ...clusterInitState, ...cluster },
    userReducer: { ...userInitState, ...user },
  };
};

const customRender = (
  ui: ReactElement,
  preloadedState: RootState = makeCustomStore(),
  { store = configureStore({ reducer: rootReducer, preloadedState }), ...renderOptions } = {},
  history = createMemoryHistory(),
): RenderResult<Queries, HTMLElement> => {
  const Wrapper: any = ({ children }: { children: ReactElement }) => {
    return (
      <Provider store={store}>
        <Router history={history}>{children}</Router>
      </Provider>
    );
  };
  return render(ui, { wrapper: Wrapper, ...renderOptions });
};

const createWithWrapper = (children: ReactElement) =>
  renderer.create(
    <Provider store={appStore}>
      <Router history={createMemoryHistory()}>{children}</Router>
    </Provider>,
  );

// re-export everything
export * from "@testing-library/react";
// override render method
export { customRender as render, createWithWrapper, makeCustomStore };
