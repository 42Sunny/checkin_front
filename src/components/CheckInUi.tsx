import { Modal, Box as MuiBox } from "@mui/material";
import React, { useEffect, useState } from "react";
import moment from "moment";
import classes from "../styles/components/CheckInUi.module.css";
import useCluster from "../utils/hooks/useCluster";
import Button from "./Button";
import Checkbox from "./Checkbox";
import List from "./List";
import Box from "./Box";

const checkLists = [
  "발열 체크시 37.5도 이하인 것을 확인했습니다.",
  "임시 출입카드 분실 시 분실 비용이 발생하는 것을 확인했습니다.",
  "마스크를 반드시 상시 착용하고 방역수칙을 준수할 것을 약속하며, 모든 설문을 이상없이 작성했음을 확인합니다.",
];

const isWeekend = () => {
  const today = new Date(moment(new Date()).local().format());
  console.log(today);

  if (today.getDay() === 6 || today.getDay() === 0) return true;
  return false;
};
const deskLunchTime = isWeekend() ? "11:00 ~ 12:00" : "13:00 ~ 14:00";

interface CardInputProps {
  cardNum: string;
  handleCardNumberChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

const CardInput: React.FC<CardInputProps> = ({ cardNum, handleCardNumberChange }) => (
  <>
    <input
      className={classes.cardNumber}
      type='number'
      min={1}
      step={1}
      value={cardNum}
      inputMode='decimal' /* 숫자형 키패드 */
      placeholder='카드 번호'
      onChange={handleCardNumberChange}
    />
  </>
);

const CheckList: React.FC = () => (
  <div className={classes["check-list-wrapper"]}>
    <ul className={classes["check-list"]}>
      {checkLists.map((checkList) => (
        <List key={checkList} text={checkList} />
      ))}
    </ul>
  </div>
);

interface IProps {
  handleCheckIn: (cardNum: string) => (e: React.FormEvent<HTMLFormElement>) => Promise<boolean>;
}

const CheckInUi: React.FC<IProps> = ({ handleCheckIn }) => {
  const [cardNum, setCardNum] = useState("");
  const [isChecked, setIsChecked] = useState(false);
  const [readySubmit, setReadySubmit] = useState(false);
  const [isOpened, setIsOpened] = useState(false);

  const {
    cluster: { officeHours },
  } = useCluster();

  const handleModalOpen = () => {
    setIsOpened(true);
  };
  const handleModalClose = () => {
    setIsOpened(false);
  };
  const handleCheck = (e: React.ChangeEvent<HTMLInputElement>) => {
    setIsChecked(e.target.checked);
  };

  const handleCardNumberChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setCardNum(e.target.value);
  };

  const checkSubmitCondition = () => {
    if (cardNum && isChecked) setReadySubmit(true);
    else setReadySubmit(false);
  };

  useEffect(() => {
    checkSubmitCondition();
    return () => {
      setReadySubmit(false);
    };
  });

  return (
    <>
      <Box>
        <p> 클러스터 운영시간: {officeHours}</p>
        <p> 인포데스크 점심시간 {deskLunchTime}</p>
      </Box>
      <form className={classes["check-in-form"]} onSubmit={handleCheckIn(cardNum)}>
        <Modal open={isOpened} onClose={handleModalClose}>
          <MuiBox>
            <CheckList />
          </MuiBox>
        </Modal>
        <CardInput cardNum={cardNum} handleCardNumberChange={handleCardNumberChange} />
        <div className={classes.checkWrap}>
          <Checkbox
            text='방역수칙에 동의하고 입장합니다.'
            checked={isChecked}
            onChange={handleCheck}
          />
          <button type='button' className={classes.checkDetail} onClick={handleModalOpen}>
            자세히
          </button>
        </div>
        <Button
          className={classes["check-in-button"]}
          type='submit'
          text='CHECK IN'
          disabled={!readySubmit}
        />
      </form>
    </>
  );
};

export default CheckInUi;
