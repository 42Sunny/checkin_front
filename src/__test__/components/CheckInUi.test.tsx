import userEvent from "@testing-library/user-event";
import CheckInUi from "../../components/CheckInUi";
import { createWithWrapper, render, screen } from "../test-utils";

describe("src/components/CheckInUi.tsx", () => {
  let handleCheckIn: any;
  beforeEach(() => {
    handleCheckIn = jest.fn();
  });
  it("matches snapshot", () => {
    const component = createWithWrapper(<CheckInUi handleCheckIn={handleCheckIn} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("when 자세히 button is clicked modal shows up", () => {
    render(<CheckInUi handleCheckIn={handleCheckIn} />);
    const infoBtn = screen.getByText("자세히");
    userEvent.click(infoBtn);
    const modal = screen.queryByRole("presentation");

    expect(modal).toBeInTheDocument();
  });
});
