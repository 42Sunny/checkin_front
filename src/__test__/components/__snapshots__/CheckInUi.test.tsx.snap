// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/CheckInUi.tsx matches snapshot 1`] = `
Array [
  <div
    className="box"
  >
    <p>
       클러스터 운영시간: 
      
    </p>
    <p>
       인포데스크 점심시간 
      11:00 ~ 12:00
    </p>
  </div>,
  <form
    className="check-in-form"
  >
    <input
      className="cardNumber"
      inputMode="decimal"
      min={1}
      onChange={[Function]}
      placeholder="카드 번호"
      step={1}
      tabIndex={0}
      type="number"
      value=""
    />
    <div
      className="checkWrap"
    >
      <label
        className="text"
      >
        <input
          checked={false}
          className="checkbox"
          onChange={[Function]}
          type="checkbox"
        />
        방역수칙에 동의하고 입장합니다.
      </label>
      <button
        className="checkDetail"
        onClick={[Function]}
        type="button"
      >
        자세히
      </button>
    </div>
    <button
      className="button green"
      color="green"
      disabled={true}
      type="submit"
    >
      CHECK IN
    </button>
  </form>,
]
`;
